src/hypothesis/searchstrategy/datetime.py

```diff
@@ -0,0 +1,96 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2017 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import datetime as dt
+
+from hypothesis.internal.conjecture import utils
+from hypothesis.searchstrategy.strategies import SearchStrategy
+
+__all__ = ['DatetimeStrategy', 'TimedeltaStrategy']
+
+
+class DatetimeStrategy(SearchStrategy):
+
+    def __init__(self, min_datetime, max_datetime, timezones_strat):
+        assert isinstance(min_datetime, dt.datetime)
+        assert isinstance(max_datetime, dt.datetime)
+        assert min_datetime.tzinfo is None
+        assert max_datetime.tzinfo is None
+        assert min_datetime <= max_datetime
+        assert isinstance(timezones_strat, SearchStrategy)
+        self.min_dt = min_datetime
+        self.max_dt = max_datetime
+        self.tz_strat = timezones_strat
+
+    def do_draw(self, data):
+        # This looks like it could loop forever, but in the worst case it
+        # would fail after exhausting the underlying entropy.  Obviously this
+        # is still bad, but even pathological bounds only loop a few times.
+        while True:
```

Comment:
!!!

I am very dubious of code that can do an infinite loop, doubly so if it touches datetime code. (The whole point of this strategy is that datetimes are fiddly and hard and have weird edge cases, so I remain not totally convinced that this does terminate cleanly.)

I’d rather have this be a `for` loop with a finite number of iterations that drops a healthcheck warning if we fall off the end of the loop. No risk of an infinite loop, and we get a better idea of how many times we expect this to take.